js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,385 @@
+					continue
+				}
+				for _, key := range headersObj.Keys() {
+					header.Set(key, headersObj.Get(key).String())
+				}
+			case "tags":
+				tagsV := params.Get(k)
+				if goja.IsUndefined(tagsV) || goja.IsNull(tagsV) {
+					continue
+				}
+				tagObj := tagsV.ToObject(rt)
+				if tagObj == nil {
+					continue
+				}
+				for _, key := range tagObj.Keys() {
+					tags[key] = tagObj.Get(key).String()
+				}
+			}
+		}
+
+	}
+
+	// Pass a custom net.Dial function to websocket.Dialer that will substitute
+	// the underlying net.Conn with our own tracked netext.Conn
+	netDial := func(network, address string) (net.Conn, error) {
+		return state.Dialer.DialContext(ctx, network, address)
+	}
+
+	wsd := websocket.Dialer{
+		NetDial: netDial,
+		Proxy:   http.ProxyFromEnvironment,
+	}
+
+	start := time.Now()
+	conn, response, connErr := wsd.Dial(url, header)
+	connectionEnd := time.Now()
+	connectionDuration := stats.D(connectionEnd.Sub(start))
+
+	socket := Socket{
+		ctx:                ctx,
+		conn:               conn,
+		eventHandlers:      make(map[string][]goja.Callable),
+		pingSendTimestamps: make(map[string]time.Time),
+		scheduled:          make(chan goja.Callable),
+		done:               make(chan struct{}),
+	}
+
+	// Run the user-provided set up function
+	setupFn(goja.Undefined(), rt.ToValue(&socket))
+
+	if connErr != nil {
+		// Pass the error to the user script before exiting immediately
+		socket.handleEvent("error", rt.ToValue(connErr))
+		return response
+	}
+	defer conn.Close()
+
+	tags["status"] = strconv.Itoa(response.StatusCode)
+	tags["subprotocol"] = response.Header.Get("Sec-WebSocket-Protocol")
+
+	// The connection is now open, emit the event
+	socket.handleEvent("open")
+
+	// Pass ping/pong events through the main control loop
+	pingChan := make(chan string)
+	pongChan := make(chan string)
+	conn.SetPingHandler(func(msg string) error { pingChan <- msg; return nil })
+	conn.SetPongHandler(func(pingID string) error { pongChan <- pingID; return nil })
+
+	readDataChan := make(chan []byte)
+	readErrChan := make(chan error)
+
+	// Wraps a couple of channels around conn.ReadMessage
+	go readPump(conn, readDataChan, readErrChan)
+
+	// This is the main control loop. All JS code (including error handlers)
+	// should only be executed by this thread to avoid race conditions
+	for {
+		select {
+		case <-pingChan:
+			// Handle pings received from the server
+			socket.handleEvent("ping")
+
+		case pingID := <-pongChan:
+			// Handle pong responses to our pings
+			socket.trackPong(pingID)
+			socket.handleEvent("pong")
+
+		case readData := <-readDataChan:
+			socket.msgsReceived++
+			socket.handleEvent("message", rt.ToValue(string(readData)))
+
+		case readErr := <-readErrChan:
+			socket.handleEvent("error", rt.ToValue(readErr))
+
+		case scheduledFn := <-socket.scheduled:
+			scheduledFn(goja.Undefined())
+
+		case <-ctx.Done():
+			// This means that K6 is shutting down (e.g., during an interrupt)
```

Comment:
Actually, it means that the VU is shutting down - the VU may just be going to sleep because the test was scaled down.